### ISSUE TITLE
Single config for ml-host/ml-port

### ISSUE THREAD

**Issue by grtjn** [marklogic/slush-marklogic-node#151]
In a format that works for Roxy, gulp, and potentially others in future?

**Comment by grtjn** [comment]
Even better would be something that allows defaults, and env-specific overrides, like Roxy..

**Comment by grtjn** [comment]
Considering patching Roxy using app_specific.rb. Simplest way of reading JSON (using core gems):

```
require 'yaml'
items = YAML.load_file('local.json')
items.each do |key, item|
  puts "#{key}:#{item}"
end
```

And then we can start considering top-level json configs like defaults.json, and local/dev/prod.json, containing something like:

```
{
  "ml-host": "localhost",
  "ml-port": 8040,
  "app-port": 9040
}
```

And potentially a few other things that would be convenient for developers to have readily available..

**Comment by hunterwilliams** [comment]
So I want to discuss this a tiny bit before jumping in.

Currently ports live in:

etc/init.d/node-express-service - Which is a bash script to start it as a service via forever
gulp.config.js - config JSON used by Gulp
deploy/build.properties - used for Roxy

I'm not totally worried about the bash script as it's probably only going to be ran in production or higher envs where a property override wouldn't exist. So that leaves gulp.config.js and build.properties.


I think we need to wait for Roxy to support JSON
* Is Roxy planning to add JSON support? 
  * If so I don't think slush should currently facilitate its work as you won't have overrides working properly without some hacks.
  * If it does or will should it just generate the json for gulp during the slush build? If so, then we just need to add whatever lines it doesn't support. **IDEAL**
* Or should gulp during the server command figure out what build.properties or read from them?
  * This seems like a bad move because we'd be tightly coupled to replicate however roxy does it's properties file decisions. 

Basically I think until Roxy supports building JSON properties this should wait otherwise you'll just get weird behavior. I think changing the ports in 2 places isn't a terrible concern as *most* developers probably will not need to change them as much since now you can do pick different ones during generation.

If it already supports JSON my mistake and I'll implement next week

**Comment by grtjn** [comment]
Anything inside `etc/` is for server deployments only, though you could consider deployment in a local vm a server deployment too. I usually run the node part on the host, and sometimes have ML in a vm. (Depends mostly on which version I need..) So, I agree, and also lean towards ignoring `etc/` in this discussion.

To continue on that I think the slush generator should only ask settings for the `local` environment, specific for the user that happens to run the generator. You can't expect an SE to know which ports will be available on the final server, nor assume all developers will use the same ports.

So, instead of doing global replaces everywhere, I would prefer to write local properties instead (whatever shape, if necessary multiple places), that will override the local(host) behavior for the entire stack. These local properties should be in the git ignore list, so they will not be shared with colleagues.

Regarding Roxy: instead of touching `deploy/build.properties`, we could simply generate a `deploy/local.properties`. That is how Roxy has operated from the start. Fastest way out here.

Regarding Gulp/node: there is no such notion yet of a local properties of any kind. We would have to introduce that ourselves. And I might have a suggestion for that. How about renaming the `gulp serve-*` to `serve-local`, `serve-dev`, and `serve-prod`. The current serve-dev would become the serve-local. The current serve-dist would split into serve-dev and serve-prod. And all three would read settings as they did before, but also look for a `/[local | dev | prod].json`. Command-line param is strongest, ENV next, env properties then, gulp.config.js/gulpfile.js last.

Regarding making Roxy read that /[env].json as well, I can look into that. I found a simple way to read JSON in Roxy, and I think we can override its built-in behavior by patching `deploy/app_specific.rb`. I can do that in a separate PR..

**Comment by hunterwilliams** [comment]
Makes sense I think. I'll have to toss all my fancy regex. This being said should I remove modifying the forever script?

**Comment by grtjn** [comment]
Well, we could do both...

How about asking explicitly for prod/default, and local settings? prod/default is for the regex stuff you have now, the local settings go into the local env file..

**Comment by patrickmcelwee** [comment]
Seems to me that throwing away fancy regex almost always means you are on the right track ;-). On the assumption that it is not a frequent need, I favor not asking explicitly about prod/defaults, and instead documenting how to change them if necessary.

**Comment by grtjn** [comment]
Fair enough. Add to the list for cookbook then? :)

**Comment by hunterwilliams** [comment]
I was totally kidding about keeping the regex

**Comment by hunterwilliams** [comment]
Not getting to today

**Comment by hunterwilliams** [comment]
So while this is going on I want to point out this section of gulpfile.js

```javascript
function getNodeOptions(isDev) {
  var port = args['app-port'] || process.env.PORT || config.defaultPort || 9070;
  return {
    script: config.nodeServer,
    delayTime: 1,
    env: {
      'PORT': port,
      'NODE_ENV': isDev ? 'dev' : 'build',
      'APP_PORT': port,
      'ML_HOST': args['ml-host'] || process.env.ML_HOST || config.marklogic.host || 'localhost',
      'ML_PORT': args['ml-port'] || process.env.ML_PORT || config.marklogic.port || 8040
    },
    watch: [config.server]
  };
}
```

Should this be slid over to gulp-config.js? However honestly the whole gulp-config.js situation isn't honestly much better.

**Comment by grtjn** [comment]
Yeah, it looked a bit odd having that in gulpfile.js itself. But we can also rearrange once we have settled on the functional side of this ticket..

**Comment by grtjn** [comment]
gulp-{env}.json working now! Up next, monkey patching Roxy to read that as well, but I will create a separate ticket for that, and assign to future milestone. It can wait..
